File: java/holiday-sdk-j25/src/main/java/com/github/gmkits/holiday/sdk/j25/DayInfo.java
package com.github.gmkits.holiday.sdk.j25;

import java.util.List;
import java.util.Map;

/**
 * 单日节假日信息。
 *
 * <p>SDK 内部 DTO，与服务端 {@code DayInfo} 字段对齐；保留 {@code holiday}/{@code workday}
 * 等服务端旧布尔字段，由 {@link HolidayClient} 解析时填充到对应不可变字段。</p>
 *
 * @param date               日期（{@code YYYY-MM-DD}）
 * @param regionCode         地区代码
 * @param calendarSystem     历法（{@code GREGORIAN} / {@code CHINESE_LUNAR}）
 * @param holiday            是否节假日
 * @param workday            是否工作日
 * @param weekend            是否周末
 * @param statutoryHoliday   是否法定节假日
 * @param adjustedWorkday    是否调班工作日
 * @param holidayNames       多语言节假日名称（locale → names）
 * @param labels             标签
 * @param sourceVersion      数据来源版本
 * @param extensions         扩展字段（lunar / solarTerm 等）
 */
public record DayInfo(
        String date,
        String regionCode,
        String calendarSystem,
        boolean holiday,
        boolean workday,
        boolean weekend,
        boolean statutoryHoliday,
        boolean adjustedWorkday,
        Map<String, List<String>> holidayNames,
        List<String> labels,
        String sourceVersion,
        Map<String, Object> extensions
) {
}

File: dotnet/CnHolidayKit/DayInfo.cs
using System.Collections.ObjectModel;
using System.Linq;

namespace CnHolidayKit;

public sealed class DayInfo
{
    public const string LunarExtensionKey = "lunar";
    public const string SolarTermExtensionKey = "solarTerm";

    private static readonly IReadOnlyDictionary<string, IReadOnlyList<string>> EmptyHolidayNames =
        new ReadOnlyDictionary<string, IReadOnlyList<string>>(new Dictionary<string, IReadOnlyList<string>>(StringComparer.Ordinal));

    private static readonly IReadOnlyList<string> EmptyLabels = Array.AsReadOnly(Array.Empty<string>());

    private static readonly IReadOnlyDictionary<string, object> EmptyExtensions =
        new ReadOnlyDictionary<string, object>(new Dictionary<string, object>(StringComparer.Ordinal));

    public DayInfo(
        DateOnly date,
        string regionCode,
        CalendarSystem calendarSystem,
        bool isHoliday,
        bool isWorkday,
        bool isWeekend,
        bool isStatutoryHoliday,
        bool isAdjustedWorkday,
        IReadOnlyDictionary<string, IReadOnlyList<string>>? holidayNames = null,
        IReadOnlyList<string>? labels = null,
        string? sourceVersion = null,
        IReadOnlyDictionary<string, object>? extensions = null)
    {
        Date = date;
        RegionCode = regionCode ?? throw new ArgumentNullException(nameof(regionCode));
        CalendarSystem = calendarSystem;
        IsHoliday = isHoliday;
        IsWorkday = isWorkday;
        IsWeekend = isWeekend;
        IsStatutoryHoliday = isStatutoryHoliday;
        IsAdjustedWorkday = isAdjustedWorkday;
        HolidayNames = FreezeHolidayNames(holidayNames);
        Labels = FreezeList(labels);
        SourceVersion = sourceVersion ?? string.Empty;
        Extensions = FreezeExtensions(extensions);
    }

    public DateOnly Date { get; }

    public string RegionCode { get; }

    public CalendarSystem CalendarSystem { get; }

    public bool IsHoliday { get; }

    public bool IsOfficialHoliday => IsHoliday && HolidayNames.Count > 0;

    public bool IsWorkday { get; }

    public bool IsWeekend { get; }

    public bool IsStatutoryHoliday { get; }

    public bool IsAdjustedWorkday { get; }

    public IReadOnlyDictionary<string, IReadOnlyList<string>> HolidayNames { get; }

    public IReadOnlyList<string> Labels { get; }

    public string SourceVersion { get; }

    public IReadOnlyDictionary<string, object> Extensions { get; }

    public LunarDateInfo? LunarDate => GetExtension<LunarDateInfo>(LunarExtensionKey);

    public SolarTermInfo? SolarTerm => GetExtension<SolarTermInfo>(SolarTermExtensionKey);

    public T? GetExtension<T>(string key) where T : class
    {
        if (!Extensions.TryGetValue(key, out var value))
        {
            return null;
        }

        return value as T;
    }

    public override string ToString()
    {
        return $"{Date:yyyy-MM-dd} [{RegionCode}] Holiday={IsHoliday}, Workday={IsWorkday}";
    }

    private static IReadOnlyDictionary<string, IReadOnlyList<string>> FreezeHolidayNames(
        IReadOnlyDictionary<string, IReadOnlyList<string>>? holidayNames)
    {
        if (holidayNames is null || holidayNames.Count == 0)
        {
            return EmptyHolidayNames;
        }

        var copy = new Dictionary<string, IReadOnlyList<string>>(holidayNames.Count, StringComparer.Ordinal);
        foreach (var pair in holidayNames)
        {
            copy[pair.Key] = FreezeList(pair.Value);
        }

        return new ReadOnlyDictionary<string, IReadOnlyList<string>>(copy);
    }

    private static IReadOnlyList<string> FreezeList(IEnumerable<string>? values)
    {
        if (values is null)
        {
            return EmptyLabels;
        }

        var materialized = values as string[] ?? values.ToArray();
        if (materialized.Length == 0)
        {
            return EmptyLabels;
        }

        return Array.AsReadOnly(materialized.ToArray());
    }

    private static IReadOnlyDictionary<string, object> FreezeExtensions(
        IReadOnlyDictionary<string, object>? extensions)
    {
        if (extensions is null || extensions.Count == 0)
        {
            return EmptyExtensions;
        }

        return new ReadOnlyDictionary<string, object>(new Dictionary<string, object>(extensions, StringComparer.Ordinal));
    }
}
